pymedphys/_experimental/pseudonymisation/strategy.py

```diff
@@ -0,0 +1,379 @@
+
+    return replacement_value
+
+
+def _pseudonymise_plaintext(value):
+    """
+    Appropriate for DICOM VR that are unstructured text, e.g. SH, AE, LO
+    Not appropriate for Numeric Strings (DS, IS)
+    Not appropriate for DateTime (DA, TM, DT)
+
+    Parameters
+    ----------
+    value : String
+        the DICOM element value for an element of type SH, AE, LO.
+
+    Returns
+    -------
+    The base64 text representation of the hashed input value
+
+    """
+    try:
+        encoded_value = value.encode("ASCII")
+    except AttributeError:
+        encoded_value = str(value).encode("ASCII")
+
+    my_hash_func = hashlib.new("sha3_256")
+    my_hash_func.update(encoded_value)
+    my_digest = my_hash_func.digest()
+    # my_hex_digest = HASH3_256.hexdigest()
+    # print ( "Hex: " + my_hex_digest )
+    # print ( "Base64: " + str(base64.standard_b64encode(my_digest)) )
+    my_pseudonym = base64.standard_b64encode(my_digest)
+    text_pseudonym = my_pseudonym.decode("ASCII")
+    # eliminate trailing '='
+    pseudonym_length = len(text_pseudonym)
+    if text_pseudonym[pseudonym_length - 1] == "=":
+        text_pseudonym = text_pseudonym[0 : pseudonym_length - 1]
+    return text_pseudonym
+
+
+def _strip_plus_slash_from_base64(value):
+    if value is None:
+        return None
+    return value.replace("+", "").replace("/", "")
+
+
+def _pseudonymise_AE(value):
+    my_pseudonym = _pseudonymise_plaintext(value)
+    my_sliced_pseudonym = my_pseudonym[0:16]
+    return my_sliced_pseudonym
+
+
+def _pseudonymise_AS(value):
+    """
+    Attempt to provide an Age that is similar but not a match
+    Round down to 80 if over 80
+    Round down to decade if 20 to 80
+    Add a pseudo-random value from -2 to 2 if 10 to 20
+    If less than 10, switch to smaller unit Y->M->W->D and add a
+    pseudo-random value from -2 to 2 to the value after converting to the
+    smaller unit.
+
+    Parameters
+    ----------
+    value : string conforming to DICOM AS VR
+        DESCRIPTION.
+
+    Returns
+    -------
+    somewhat pseudonymised equivalent of the age as DICOM VR AS
+
+    """
+    random.seed()
+    increment = random.randrange(-2, 2)
+    numeric = int(value[0:3])
+    unit = value[3]
+    pseudo_unit = unit
+    if numeric > 20:
+        pseudo_numeric = numeric - (numeric % 10)
+        if numeric > 80:
+            pseudo_numeric = 80  # very old people are rare, too rare to avoid id
+    elif numeric > 10:
+        pseudo_numeric = numeric + increment
+    else:  # highly problematic if the value is too small
+        if unit == "Y":
+            pseudo_numeric = 12 * numeric + increment
+            pseudo_unit = "M"
+        elif unit == "M":
+            pseudo_numeric = 4 * numeric + increment
+            pseudo_unit = "W"
+        elif unit == "W":
+            pseudo_numeric = 7 * numeric + increment
+            pseudo_unit = "D"
+
+    pseudo_age = str(pseudo_numeric).zfill(3) + pseudo_unit
```

Comment:
Do we want this to be reversible?

Something that would be reversible, but also unable to be determined without a key (the large prime number) would be modulo arithmetic with a modulus that is prime and multiplying by a very large number that is also prime, and then offset so as to always be older than 20 say.

Reply:
No.  As above, if someone needs the PHI, they should already have it with the original data.
If they need to find their way back to that original data see previous comment/response.